File: Backend/controllers/sensorController.js
const fs = require('fs');
const path = require('path');

const DATA_MASUK_DIR = path.join(__dirname, '..', '..', 'data_masuk');
const INGEST_TOKEN = process.env.SENSOR_INGEST_TOKEN || '';

const pad = (n) => String(n).padStart(2, '0');
const nowTimestamp = () => {
  const d = new Date();
  return `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())} ${pad(d.getHours())}:${pad(d.getMinutes())}:${pad(d.getSeconds())}`;
};

const parseTimestamp = (ts) => {
  if (typeof ts === 'string' && /^\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}$/.test(ts.trim())) {
    return ts.trim();
  }
  return nowTimestamp();
};

const ingest = (req, res) => {
  const auth = req.headers['authorization'] || '';
  if (!INGEST_TOKEN || auth !== `Bearer ${INGEST_TOKEN}`) {
    return res.status(401).json({ success: false, message: 'Token tidak valid' });
  }

  const { kja_id, ph, salinitas, suhu, kekeruhan, status, timestamp } = req.body;
  const nilai = [ph, salinitas, suhu, kekeruhan].map(Number);
  if (kja_id === undefined || nilai.some((v) => isNaN(v))) {
    return res.status(400).json({ success: false, message: 'Data tidak lengkap' });
  }

  const ts = parseTimestamp(timestamp);
  try {
    if (!fs.existsSync(DATA_MASUK_DIR)) fs.mkdirSync(DATA_MASUK_DIR, { recursive: true });
    const file = path.join(DATA_MASUK_DIR, `sensor_${ts.slice(0, 10)}.csv`);
    if (!fs.existsSync(file)) {
      fs.writeFileSync(file, 'kja_id,ph,salinitas,suhu,kekeruhan,status,timestamp\n');
    }
    fs.appendFileSync(
      file,
      `${kja_id},${nilai[0].toFixed(2)},${nilai[1].toFixed(2)},${nilai[2].toFixed(2)},${nilai[3].toFixed(2)},${status || ''},${ts}\n`
    );
    return res.status(201).json({ success: true, message: 'Data tersimpan' });
  } catch (err) {
    console.error('Gagal menulis CSV data_masuk:', err.message);
    return res.status(500).json({ success: false, message: 'Gagal menyimpan data' });
  }
};

module.exports = { ingest };

File: backend_python/controllers/sensor_controller.py
import json
import os
import re
from datetime import datetime

from flask import request
from dotenv import load_dotenv

from config.database import BASE_DIR
from utils.helpers import fail, ok

load_dotenv(os.path.join(BASE_DIR, ".env"))
INGEST_TOKEN = os.getenv("SENSOR_INGEST_TOKEN", "")
DATA_MASUK_DIR = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "data_masuk")
HEADER = ["kja_id", "ph", "salinitas", "suhu", "kekeruhan", "status", "timestamp"]
TS_PATTERN = re.compile(r"^\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}$")


def _parse_timestamp(ts):
    if isinstance(ts, str) and TS_PATTERN.match(ts.strip()):
        return ts.strip()
    return datetime.now().strftime("%Y-%m-%d %H:%M:%S")


def ingest():
    auth = request.headers.get("Authorization", "")
    if not INGEST_TOKEN or auth != f"Bearer {INGEST_TOKEN}":
        return fail("Token tidak valid", 401)

    body = request.get_json(silent=True) or {}
    nilai = []
    for k in ("ph", "salinitas", "suhu", "kekeruhan"):
        try:
            v = float(body.get(k))
            if v != v:
                raise ValueError
            nilai.append(f"{v:.2f}")
        except (TypeError, ValueError):
            return fail(f"Field {k} tidak valid", 400)
    if body.get("kja_id") is None:
        return fail("kja_id wajib ada", 400)

    ts = _parse_timestamp(body.get("timestamp"))
    os.makedirs(DATA_MASUK_DIR, exist_ok=True)
    path = os.path.join(DATA_MASUK_DIR, f"sensor_{ts[:10]}.csv")
    baru = not os.path.exists(path)
    with open(path, "a", encoding="utf-8") as f:
        if baru:
            f.write(",".join(HEADER) + "\n")
        f.write(
            f"{body['kja_id']},{nilai[0]},{nilai[1]},{nilai[2]},{nilai[3]},{body.get('status', '')},{ts}\n"
        )
    row = {
        "kja_id": body["kja_id"],
        "ph": nilai[0],
        "salinitas": nilai[1],
        "suhu": nilai[2],
        "kekeruhan": nilai[3],
        "status": str(body.get("status", "")),
        "timestamp": ts,
    }
    print(f"[DATA] kja_id={row['kja_id']} -> {os.path.basename(path)}")
    payload = {"success": True, "message": "Data tersimpan", "data": row}
    from flask import jsonify

    return jsonify(payload), 201
